src/tokenizer/shell.cpp

```diff
@@ -0,0 +1,526 @@
+        break;
+    case shell_token_type::backtick_substitution_close:
+        os << "backtick_substitution_close";
+        break;
+    case shell_token_type::dollar:
+        os << "dollar";
+        break;
+    case shell_token_type::redirection:
+        os << "redirection";
+        break;
+    case shell_token_type::command_substitution_open:
+        os << "command_substitution_open";
+        break;
+    case shell_token_type::command_substitution_close:
+        os << "command_substitution_close";
+        break;
+    case shell_token_type::parenthesis_open:
+        os << "parenthesis_open";
+        break;
+    case shell_token_type::parenthesis_close:
+        os << "parenthesis_close";
+        break;
+    case shell_token_type::curly_brace_open:
+        os << "curly_brace_open";
+        break;
+    case shell_token_type::curly_brace_close:
+        os << "curly_brace_close";
+        break;
+    case shell_token_type::process_substitution_open:
+        os << "process_substitution_open";
+        break;
+    case shell_token_type::process_substitution_close:
+        os << "process_substitution_close";
+        break;
+    case shell_token_type::subshell_open:
+        os << "subshell_open";
+        break;
+    case shell_token_type::subshell_close:
+        os << "subshell_close";
+        break;
+    case shell_token_type::compound_command_open:
+        os << "compound_command_open";
+        break;
+    case shell_token_type::compound_command_close:
+        os << "compound_command_close";
+        break;
+    }
+    return os;
+}
+
+shell_tokenizer::shell_tokenizer(
+    std::string_view str, std::unordered_set<shell_token_type> skip_tokens)
+    : base_tokenizer(str, std::move(skip_tokens))
+{
+    shell_scope_stack_.reserve(8);
+
+    if (!redirection_regex.ok()) {
+        throw std::runtime_error("redirection regex not valid: " + redirection_regex.error_arg());
+    }
+}
+
+void shell_tokenizer::tokenize_delimited_token(std::string_view delimiter, shell_token_type type)
+{
+    shell_token token;
+    token.index = index();
+    token.type = type;
+
+    std::size_t idx = 0;
+    while (idx < delimiter.size() && advance()) {
+        idx += static_cast<int>(peek() == delimiter[idx]);
+    }
+
+    token.str = substr(token.index, index() - token.index + 1);
+    emplace_token(token);
+}
+
+void shell_tokenizer::tokenize_variable()
+{
+    shell_token token;
+    token.index = index();
+    token.type = shell_token_type::variable;
+
+    // Skip dollar
+    advance();
+
+    // We know the first character is $
+    auto c = peek();
+    if (c == '-' || c == '?' || c == '@' || c == '#' || c == '*' || c == '$' || c == '!' ||
+        ddwaf::isdigit(c)) {
+        // Special variable, these are one character long + dollar
+        token.str = substr(token.index, 2);
+    } else if (c == '{') {
+        while (advance() && peek() != '}') {}
+        token.str = substr(token.index, index() - token.index + 1);
```

Comment:
Fixed, somewhat. Variables are a bit challenging so there are some cases which aren't parsed adequately but we can work on fixing those later on as this could go on ad infinitum with this language.